seatunnel-connectors-v2/connector-iotdb-v2/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdbv2/source/IoTDBv2SourceSplitEnumerator.java

```diff
@@ -0,0 +1,261 @@
+            sqlBase = sqls[0];
+            sqlAlign = sqls[1];
+        }
+        sqls = sqlBase.split("(?i)" + SQL_WHERE);
+        if (sqls.length > SQL_WHERE_SPLIT_LENGTH) {
+            throw new IotdbConnectorException(
+                    CommonErrorCodeDeprecated.ILLEGAL_ARGUMENT,
+                    "sql should not contain more than one where");
+        }
+        if (sqls.length > 1) {
+            sqlBase = sqls[0];
+            sqlCondition = sqls[1];
+        }
+        long size = (end - start) / numPartitions + 1;
+        long remainder = (end + 1 - start) % numPartitions;
+        if (end - start < numPartitions) {
+            numPartitions = (int) (end - start);
+        }
+        long currentStart = start;
+        int i = 0;
+        while (i < numPartitions) {
+            String query =
+                    " where ("
+                            + RESERVED_TIME
+                            + " >= "
+                            + currentStart
+                            + " and "
+                            + RESERVED_TIME
+                            + " < "
+                            + (currentStart + size)
+                            + ") ";
+            i++;
+            currentStart += size;
+            if (i + 1 <= numPartitions) {
+                currentStart = currentStart - remainder;
+            }
+            query = sqlBase + query;
+            if (!Strings.isNullOrEmpty(sqlCondition)) {
+                query = query + " and ( " + sqlCondition + " ) ";
+            }
+            if (!Strings.isNullOrEmpty(sqlAlign)) {
+                query = query + " align by " + sqlAlign;
+            }
+            iotDBSourceSplits.add(new IoTDBv2SourceSplit(String.valueOf(query.hashCode()), query));
+        }
+        return iotDBSourceSplits;
+    }
+
+    @Override
+    public void addSplitsBack(List<IoTDBv2SourceSplit> splits, int subtaskId) {
+        log.debug("Add back splits {} to IoTDBSourceSplitEnumerator.", splits);
+        if (!splits.isEmpty()) {
+            addPendingSplit(splits);
+            assignSplit(Collections.singletonList(subtaskId));
+        }
+    }
+
+    @Override
+    public int currentUnassignedSplitSize() {
+        return pendingSplit.size();
+    }
+
+    @Override
+    public void registerReader(int subtaskId) {
+        log.debug("Register reader {} to IoTDBSourceSplitEnumerator.", subtaskId);
+        if (!pendingSplit.isEmpty()) {
+            assignSplit(Collections.singletonList(subtaskId));
+        }
+    }
+
+    private void addPendingSplit(Collection<IoTDBv2SourceSplit> splits) {
+        int readerCount = context.currentParallelism();
+        for (IoTDBv2SourceSplit split : splits) {
+            int ownerReader = getSplitOwner(split.splitId(), readerCount);
+            log.info("Assigning {} to {} reader.", split, ownerReader);
+            pendingSplit.computeIfAbsent(ownerReader, r -> new ArrayList<>()).add(split);
+        }
```

Comment:
`pendingSplit` is a competitive resource and should be locked，
synchronized (stateLock) {
...
}